feat: Dependency tree command (#1069)
prefix-dev/pixi: docs/cli.md

 xz                5.2.6       h166bdaf_0          408.6 KiB  conda  xz-5.2.6-h166bdaf_0.tar.bz2
```

## `tree`

Display the project's packages in a tree. Highlighted packages are those specified in the manifest.

The package tree can also be inverted (`-i`), to see which packages require a specific dependencies.

##### Arguments

- `REGEX` optional regex of which direct dependencies to filter the tree to, or which dependencies to start with when inverting the tree.

##### Options

- `--invert (-i)`: Invert the dependency tree, that is given a `REGEX` pattern that matches some packages, show all the packages that depend on those.
- `--platform <PLATFORM> (-p)`: The platform to list packages for. Defaults to the current platform
- `--manifest-path <MANIFEST_PATH>`: The path to [manifest file](configuration.md), by default it searches for one in the parent directories.

> - `--environment`(`-e`): The environment's packages to list, if non is provided the default environment's packages will be listed.

Reviewer: ```suggestion
- `--environment (-e)`: The environment's packages to list, if non is provided the default environment's packages will be listed.
```

we should also make this consistent in `list` (and all other occurrences that i didn't spot)

- `--frozen`: Install the environment as defined in the lockfile. Without checking the status of the lockfile. It can also be controlled by the `PIXI_FROZEN` environment variable (example: `PIXI_FROZEN=true`).
- `--locked`: Only install if the `pixi.lock` is up-to-date with the [manifest file](configuration.md)[^1]. It can also be controlled by the `PIXI_LOCKED` environment variable (example: `PIXI_LOCKED=true`). Conflicts with `--frozen`.
- `--no-install`: Don't install the environment for pypi solving, only update the lock-file if it can solve without installing. (Implied by `--frozen` and `--locked`)

```shell
pixi tree
pixi tree pre-commit
pixi tree -i yaml
pixi tree --environment docs
```

!!! warning
    Use `-v` to show which `pypi` packages are not yet parsed correctly. The `extras` and `markers` parsing is still under development.

Output will look like this, where direct packages in the [manifest file](configuration.md) will be green.

> Once a package has been displayed once, the tree won't continue to recurse through it's dependencies (compare the first time `livzlib` appears, vs the rest), and it will instead be marked with a star `(*)`.

Reviewer: ```suggestion
Once a package has been displayed once, the tree won't continue to recurse through its dependencies (compare the first time `libzlib` appears, vs the rest), and it will instead be marked with a star `(*)`.
```

not sure `libzlibz is the best example for this since it doesn't require any dependencies 😅

Author: Ya, it's not great, but after collapsing things it was the first one to show up a second time. I'll play with snipping the output in different spots and see if something else shows up better.

##### Options

- `--invert (-i)`: Invert the dependency tree, that is given a `REGEX` pattern that matches some packages, show all the packages that depend on those.
- `--platform <PLATFORM> (-p)`: The platform to list packages for. Defaults to the current platform
- `--manifest-path <MANIFEST_PATH>`: The path to [manifest file](configuration.md), by default it searches for one in the parent directories.
- `--environment`(`-e`): The environment's packages to list, if non is provided the default environment's packages will be listed.
- `--frozen`: Install the environment as defined in the lockfile. Without checking the status of the lockfile. It can also be controlled by the `PIXI_FROZEN` environment variable (example: `PIXI_FROZEN=true`).
- `--locked`: Only install if the `pixi.lock` is up-to-date with the [manifest file](configuration.md)[^1]. It can also be controlled by the `PIXI_LOCKED` environment variable (example: `PIXI_LOCKED=true`). Conflicts with `--frozen`.
- `--no-install`: Don't install the environment for pypi solving, only update the lock-file if it can solve without installing. (Implied by `--frozen` and `--locked`)

> ```shell

Reviewer 2: ```suggestion
```
This line is not needed.

Author: Helps if I look at how things are being rendered by in the `list` docs, rather than blindly copying them. I've removed both of them.

Reviewer 2: Thanks!

```shell
pixi tree
pixi tree pre-commit
pixi tree -i yaml
pixi tree --environment docs
```

!!! warning

>     `pixi tree`'s support for PyPI packages is still under development, and not all packages are currently being parsed. Add `-v` to increase verbosity to display notices to see which packages are currently unable to be parsed.

Reviewer 2: I would write a warning in a more concrete way with a call to action upfront to make it easier to parse. 
That said like my other comment, I don't completely get the warning. Could you elaborate some more?
```suggestion
    Use `-v` to show which `pypi` packages are not yet parsed correctly. The `extras` and `markers` parsing is still under development.
```

│   │   │   ├── pysocks v1.7.1
│   │   │   │   └── python v3.12.2 (*)
│   │   │   ├── python v3.12.2 (*)
│   │   │   └── brotli-python v1.1.0
│   │   │       ├── python v3.12.2 (*)
│   │   │       ├── python_abi v3.12 (*)
│   │   │       └── libcxx v16.0.6 (*)
│   │   ├── certifi v2024.2.2
│   │   │   └── python v3.12.2 (*)
│   │   └── charset-normalizer v3.3.2
│   │       └── python v3.12.2 (*)
│   ├── click v8.1.7
│   │   └── python v3.12.2 (*)
│   ├── pykwalify v1.8.0
│   │   ├── python v3.12.2 (*)
│   │   ├── docopt v0.6.2
│   │   ├── python-dateutil v2.9.0
│   │   │   ├── python v3.12.2 (*)
│   │   │   └── six v1.16.0 (*)
│   │   └── ruamel.yaml v0.18.6 (*)
│   └── python v3.12.2 (*)
└── tbump v6.9.0
    ├── cli-ui v0.17.2
    │   ├── colorama v0.4.6
    │   │   └── python v3.12.2 (*)
    │   ├── python v3.12.2 (*)
    │   ├── tabulate v0.9.0
    │   │   └── python v3.12.2 (*)
    │   └── unidecode v1.3.8
    │       └── python v3.12.2 (*)
    ├── python v3.12.2 (*)
    ├── schema v0.7.5
    │   ├── contextlib2 v21.6.0
    │   │   └── python v3.12.2 (*)
    │   └── python v3.12.2 (*)
    ├── tomlkit v0.12.4
    │   └── python v3.12.2 (*)
    └── docopt v0.6.2 (*)

> ```

Reviewer 2: I think the current ~120 lines could be shrunk to 20. As here it would just give an idea to the user what it could look like. 

Showing interesting things like the 4 layers deep, the `(*)` and the

A regex pattern can be specified to filter the tree to just those that show a specific direct dependency:

```shell
➜ pixi tree pre-commit
└── pre-commit v3.3.3
    ├── virtualenv v20.25.1
    │   ├── filelock v3.13.1
    │   │   └── python v3.12.2
    │   │       ├── libexpat v2.6.2
    │   │       ├── readline v8.2
    │   │       │   └── ncurses v6.4.20240210
    │   │       ├── libsqlite v3.45.2
    │   │       │   └── libzlib v1.2.13
    │   │       ├── bzip2 v1.0.8
    │   │       ├── libzlib v1.2.13 (*)
    │   │       ├── libffi v3.4.2
    │   │       ├── tk v8.6.13
    │   │       │   └── libzlib v1.2.13 (*)
    │   │       ├── xz v5.2.6
    │   │       ├── ncurses v6.4.20240210 (*)
    │   │       └── openssl v3.2.1
    │   ├── platformdirs v4.2.0
    │   │   └── python v3.12.2 (*)
    │   ├── distlib v0.3.8
    │   │   └── python v3.12.2 (*)
    │   └── python v3.12.2 (*)
    ├── pyyaml v6.0.1
    │   ├── python_abi v3.12
    │   ├── python v3.12.2 (*)
    │   └── yaml v0.2.5
    ├── nodeenv v1.8.0
    │   └── python v3.12.2 (*)
    ├── python v3.12.2 (*)
    ├── cfgv v3.3.1
    │   └── python v3.12.2 (*)
    └── identify v2.5.35

>         └── python v3.12.2 (*)

Reviewer 2: Again, would shrink this so maybe 10 or so. Using `...` as a common insert to show there has been a cut off in the text.

Author: Snipped, but I went a little longer to show marking and tree depth.

```shell
➜ pixi tree
├── pre-commit v3.3.3
│   ├── cfgv v3.3.1
│   │   └── python v3.12.2
│   │       ├── libsqlite v3.45.2
│   │       │   └── libzlib v1.2.13
│   │       ├── libzlib v1.2.13 (*)
│   │       ├── libffi v3.4.2
│   │       ├── libexpat v2.6.2
│   │       ├── readline v8.2
│   │       │   └── ncurses v6.4.20240210
│   │       ├── xz v5.2.6
│   │       ├── openssl v3.2.1
│   │       ├── bzip2 v1.0.8
│   │       ├── tk v8.6.13
│   │       │   └── libzlib v1.2.13 (*)
│   │       └── ncurses v6.4.20240210 (*)
│   ├── pyyaml v6.0.1
...
├── rust v1.76.0
│   └── rust-std-aarch64-apple-darwin v1.76.0
├── openssl v3.2.1
├── pkg-config v0.29.2
│   ├── libglib v2.78.4
│   │   ├── libiconv v1.17
│   │   ├── gettext v0.21.1
│   │   │   └── libiconv v1.17 (*)
...
```

```

> Additionly the tree can be inverted, and it can show which packages depend on a regex pattern. The packages specified in the manifest will also be highlighted (in this case `cffconvert` and `pre-commit` would be).

Reviewer 2: ```suggestion
Additionally the tree can be inverted, and it can show which packages depend on a regex pattern.
The packages specified in the manifest will also be highlighted (in this case `cffconvert` and `pre-commit` would be).
```

We like to use single sentence per line.
Could you add some new lines?

Author: Done
